### PR TITLE
New documentation for order/cart/payment best practices

### PR DESCRIPTION
**Pull request by JudeNiroshan**
New documentation for best practices section.

#### Note to reviewer
Few files have changed because of doctoc regeneration. Files which really modified are:
- BEST_PRACTICES.md
- WebComponentsIntegrationGuide.md
- IntegrationGuide.md

### CODE REVIEW

File: docs/BEST_PRACTICES.md [hunk 1]
```diff
@@ -1,3 +1,69 @@
...
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+
+## Order creation best practices
+#### AFTER a successful payment
+##### Limitations & Solutions
+- It is possible that shop was not reachable for network issues. In that case there will be a successful payment but no order.
+
+  ###### Possible solution:
+  Create an order based on transaction state changes (commercetools messages or subscriptions)
```
Comment (Fifciu) [inline docs/BEST_PRACTICES.md:23]: There are 2 cases where using commercetools messages or subscriptions is recommended. Would it be possible to cover this subject in the additional doc? 

I feel like it won't be that obvious for developers and they will just omit these 2 best practices or only the minority will do it properly and it will be very time-demanding for them.

I totally understand it might be too much for you. However, I think it will increase developer experience and nothing will prevent developers to apply every best practice then.

**Review by Fifciu — COMMENTED** [review]
A very useful one. I am happy you've written it. I have just a one suggestion

File: docs/BEST_PRACTICES.md [hunk 2]
```diff
@@ -1,3 +1,69 @@
...
+
+------  
+- It is possible to have more than 1 valid payments on cart/order. During checkout in two tabs, two redirect payments
+(like credit card and paypal) can be created for the same cart. These two payments can be completed independently in both tabs.
+ This will lead the cart that links to two successful payments.
+  ###### Possible solution:
+  Refund the payment. (It can be based on commercetools messages or subscriptions)
+
+------  
+- It is possible to have the payment amount is lower than cart amount. During the checkout the shopper opens two tabs, one tab is for cart and
```
Comment (butenkor) [inline docs/BEST_PRACTICES.md:33]: In general I would try to separate everywhere the issue title which is `It is possible to have the payment amount is lower than cart amount` from the detailed description of it and maybe make it possible to link so that one could share a deep link as also link directly within our documentation. Also I would rephrase the title to: `The amount of the successful payment is lower than the cart amount`

Comment (butenkor) [inline docs/BEST_PRACTICES.md:33]:
```suggestion
- The amount of the successful payment is lower than the cart amount. During the checkout the shopper might navigate with two tabs, one with the cart and
```

File: docs/BEST_PRACTICES.md [hunk 3]
```diff
@@ -1,3 +1,69 @@
...
+the other tab is for payment provider. Shopper can add more items to the cart so that the cart value is not equal to the
```
Comment (butenkor) [inline docs/BEST_PRACTICES.md:34]:
```suggestion
the other tab might be already a redirected payment with fixed amount like PayPal. Shopper can add more items to the cart so that the cart value is not equal to the 
```

File: docs/BEST_PRACTICES.md [hunk 4]
```diff
@@ -1,3 +1,69 @@
...
+  Refund the payment. (It can be based on commercetools messages or subscriptions)
+
+------  
+- It is possible to have the payment amount is lower than cart amount. During the checkout the shopper opens two tabs, one tab is for cart and
+the other tab is for payment provider. Shopper can add more items to the cart so that the cart value is not equal to the 
+amount that will be paid in the payment provider tab.
+
+  ###### Possible solution:
+  After a successful payment, validate the cart to check if the cart amount matches payment amount. 
+  If not, refund the payment and let shopper to pay again.
```
Comment (butenkor) [inline docs/BEST_PRACTICES.md:39]:
```suggestion
  If not, refund/cancel the payment and ask the shopper to pay again.
```

File: docs/BEST_PRACTICES.md [hunk 5]
```diff
@@ -1,3 +1,69 @@
...
+- There will be orders which will never be paid but shown in commercetools Merchant Center dashboard as revenue - which is not true
+------  
+- Not paid orders will falsely boost the concept of “best sellers” index of its products.
+------  
+- Creating an order does not allow you to modify the original cart anymore. What happens if a shopper forgot something and would like to change the cart just after initiation of the payment?
+------  
+- How to handle vouchers that can be applied only once but the shopper decided not to finish the payment and change the cart instead? It might require a new cart based on the old cart. The creation of a new cart with the same item might be an issue if an item had only stock = 1 and it has been used in not paid order already.
+
+##### Conclusion
+Both approaches have their good and bad sides, but it is recommended creating order AFTER a successful payment is less harmful.
```
Comment (butenkor) [inline docs/BEST_PRACTICES.md:54]:
```suggestion
Both approaches have their good and bad sides, but we found out that creating an order AFTER a successful payment is less harmful.
```

File: docs/BEST_PRACTICES.md [hunk 6]
```diff
@@ -1,3 +1,69 @@
...
+
+## Payment related best practices
+
+- Payment status changes will not be managed by the Adyen Integration library. Adyen Integration is base upon the
```
Comment (butenkor) [inline docs/BEST_PRACTICES.md:58]: Maybe better:
`Currently [payment status](https://docs.commercetools.com/api/projects/payments#paymentstatus) is not maintained by the integration as status of the payment can be derived from the state of the payment transaction(s).`

File: docs/BEST_PRACTICES.md [hunk 7]
```diff
@@ -1,3 +1,69 @@
...
+## Payment related best practices
+
+- Payment status changes will not be managed by the Adyen Integration library. Adyen Integration is base upon the 
+transactions and custom fields in Commercetools platform.
+- **Never delete or un-assign** created payment objects during the checkout from the cart. 
+If required — clean up unused/obsolete payment objects by another asynchronous process instead.
+
+## Order/Cart related best practices
+
+- Order/Cart modifications have to be part of the shop. It is not the responsibility of the Adyen Integration modules.
```
Comment (butenkor) [inline docs/BEST_PRACTICES.md:65]:
```suggestion
- Order/cart modifications should be part of the front end business logic. Adyen integration will neither change the cart nor the order.
```

File: docs/BEST_PRACTICES.md (hunk 1 above)
Comment (butenkor) [reply docs/BEST_PRACTICES.md:23]: Please check if it addresses your request: https://github.com/commercetools/commercetools-adyen-integration/pull/670#discussion_r622505085

File: docs/BEST_PRACTICES.md [hunk 8]
```diff
@@ -1,3 +1,69 @@
...
+- [Payment related best practices](#payment-related-best-practices)
+- [Order/Cart related best practices](#ordercart-related-best-practices)
+- [Deployment best practices](#deployment-best-practices)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+
+## Order creation best practices
+#### AFTER a successful payment
+##### Limitations & Solutions
+- It is possible that shop was not reachable for network issues. In that case there will be a successful payment but no order.
```
[the remark this answers is not in the document]
Comment (JudeNiroshan) [reply docs/BEST_PRACTICES.md]: 5cfa6c0

File: docs/BEST_PRACTICES.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (JudeNiroshan) [reply docs/BEST_PRACTICES.md]: 59e79995d190a3aef7aaf700c8353aa96dcfb136

File: docs/BEST_PRACTICES.md [hunk 9]
```diff
@@ -1,3 +1,69 @@
...
+
+  ###### Possible solution:
+  Create an order based on transaction state changes (commercetools messages or subscriptions)
+
+------  
+- It is possible to have more than 1 valid payments on cart/order. During checkout in two tabs, two redirect payments
```
[the remark this answers is not in the document]
Comment (JudeNiroshan) [reply docs/BEST_PRACTICES.md]: a9019ab065614efe373148635eb4f72599b8ee26

File: docs/BEST_PRACTICES.md [hunk 10]
```diff
@@ -1,3 +1,69 @@
...
+(like credit card and paypal) can be created for the same cart. These two payments can be completed independently in both tabs.
+ This will lead the cart that links to two successful payments.
```
[the remark this answers is not in the document]
Comment (JudeNiroshan) [reply docs/BEST_PRACTICES.md]: f4d3f72c17901cb2a54fe9955c29379d4f88f8da

File: docs/BEST_PRACTICES.md [hunk 11]
```diff
@@ -1,3 +1,69 @@
...
+  ###### Possible solution:
+  Refund the payment. (It can be based on commercetools messages or subscriptions)
```
[the remark this answers is not in the document]
Comment (JudeNiroshan) [reply docs/BEST_PRACTICES.md]: 5f2894a5b4f7b7951e9def10285595dd2a6ba1fe

File: docs/BEST_PRACTICES.md (hunk 2 above)
Comment (JudeNiroshan) [reply docs/BEST_PRACTICES.md]: d8f39a0726026abab680a9dd9a3fec009a21e10f

Comment (JudeNiroshan) [reply docs/BEST_PRACTICES.md]: 4d2db6b6052dc169cbb943161d83196b62c3e99f

File: docs/BEST_PRACTICES.md (hunk 3 above)
Comment (JudeNiroshan) [reply docs/BEST_PRACTICES.md]: d8c04ea0b3c181e15a306742019db10588e5c7a2

File: docs/BEST_PRACTICES.md (hunk 5 above)
Comment (JudeNiroshan) [reply docs/BEST_PRACTICES.md]: 55fda2ea72757ca01d4cd82e08368dfd9f92601c

File: docs/BEST_PRACTICES.md (hunk 4 above)
Comment (JudeNiroshan) [reply docs/BEST_PRACTICES.md]: d8c04ea0b3c181e15a306742019db10588e5c7a2

File: docs/BEST_PRACTICES.md (hunk 6 above)
Comment (JudeNiroshan) [reply docs/BEST_PRACTICES.md]: 198229a6c481d22ca386cfdc44990d0032ba1939

File: docs/BEST_PRACTICES.md (hunk 7 above)
Comment (JudeNiroshan) [reply docs/BEST_PRACTICES.md]: 31555e660d9cf7c558c6068709ec66cb8dd90fcd

File: docs/BEST_PRACTICES.md (hunk 1 above)
Comment (Fifciu) [reply docs/BEST_PRACTICES.md]: Yes, now it gives me the context. Thanks!

File: docs/ContributionGuide.md [hunk 12]
```diff
@@ -1,7 +1,6 @@
 <!-- START doctoc generated TOC please keep comment here to allow auto update -->
 <!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
 **Table of Contents**  
-<!-- *generated with [DocToc](https://github.com/thlorenz/doctoc)* -->
```
[the remark this answers is not in the document]
Comment (butenkor) [reply docs/ContributionGuide.md:4]: If you want to keep them then why they are removed in this PR?

File: docs/BEST_PRACTICES.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (butenkor) [reply docs/BEST_PRACTICES.md]: I created a suggestion so you do not have to create a separate commit and link since just a click on the button above "Commit suggestion" would resolve the issue and commit it for you ;)

File: docs/BEST_PRACTICES.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (butenkor) [reply docs/BEST_PRACTICES.md]: same as above...

File: docs/BEST_PRACTICES.md (hunk 2 above)
Comment (butenkor) [reply docs/BEST_PRACTICES.md]: Thanks. I had another look at the end result here: https://github.com/commercetools/commercetools-adyen-integration/blob/d8f39a0726026abab680a9dd9a3fec009a21e10f/docs/BEST_PRACTICES.md#more-than-1-valid-payments-on-the-cartorder

maybe it would be better:
- Not to use a bullet point where there is a single point only
- Do not use grey line separator between the bullet points here: https://github.com/commercetools/commercetools-adyen-integration/blob/d8f39a0726026abab680a9dd9a3fec009a21e10f/docs/BEST_PRACTICES.md#before-a-successful-payment

File: docs/ContributionGuide.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (JudeNiroshan) [reply docs/ContributionGuide.md:4]: I think now the library is not generating this comment. There was no manual change to this file in this PR

File: docs/BEST_PRACTICES.md (hunk 2 above)
Comment (JudeNiroshan) [reply docs/BEST_PRACTICES.md]: a74dbc4

**Review by butenkor — APPROVED** [review]
Looks good - Thanks!
